shared/src/main/scala/mlscript/ConstraintSolver.scala

```
@@ -818,13 +820,23 @@ class ConstraintSolver extends NormalForms { self: Typer =>
             val newBound = (cctx._1 ::: cctx._2.reverse).foldRight(rhs)((c, ty) =>
               if (c.prov is noProv) ty else mkProxy(ty, c.prov))
             lhs.upperBounds ::= newBound // update the bound
+            lhs.lbtsc.foreach {
+              case (tsc, i) =>
+                tsc.filterUB(i, rhs)
+                if (tsc.constraints.isEmpty) reportError()
```

Should probably raise a proper error here, which should refer to the failing TSC.